rosidl_generator_py/resource/_msg.py.em

```diff
@@ -103,6 +105,30 @@ class Metaclass_@(message.structure.namespaced_type.name)(type):
 @[end for]@
     }
 
+    def __new__(cls, *args):
+        new_type = super().__new__(cls, *args)
+
+        # Ugly hack here.
+        # There are two purposes for __slots__ field of message class:
+        #   1) __slots__ field is used in ROS as a list of message members.
+        #      Number of ROS packages rely on its value.
+        #   2) defining __slots__ in Python class prevents adding __dict__
+        #      to class instances. This behvior is desirable.
+        # If #1 defined in message class body, then base class member descriptors are
+        # overriden and no longer can be used. On the other side, any __slots__ value
+        # should be defined to guaratee #2.
+        # So we define empty list in message class body and modify it after type object
+        # constructed. This is neither prohibited, nor encouraged by CPython documentation.
```

Comment:
If I understand correctly, normally `__slots__` requires subclasses to only define new slots:

> ["The action of a __slots__ declaration is not limited to the class where it is defined. __slots__ declared in parents are available in child classes. However, child subclasses will get a and __weakref__ unless they also define __slots__ (which should only contain names of any additional slots)."](https://docs.python.org/3/reference/datamodel.html#notes-on-using-slots)

This defines an empty `__slots__` attribute on the subclass prior to calling `type.__new__`, which is where cPython makes the decision to add `__dict__` or not, and where the quoted note seems to matter. It then adds to `__slots__`  afterwards with the assumption that cPython won't use it for any other purpose, but it allows users to get a list of the members of a message.

I'm hesitant because it seems not very well defined when cPython will stop using `__slots__`.

An alternative option would be to make a different API for retrieving the members of a message. That would require downstream users to change their code in Rolling, but would give us freedom to put `__slots__` in base classes. 

Another alternative would be to implement the entire class in C so that there's only one class.

Maybe yet another option would be to do monkey patching instead of a metaclass. Say there was a `class Example:`  that had an `Example.__new__` method that returned an instance of `ExampleBase`. The first time it ran it would monkey patch `ExampleBase` with the methods from `Example`.

Out of all the options, my favorite is making a new API for retrieving members of the message. @clalancette What do you think? Pinging you because it would change a (maybe unintentionally) public API without a good way to tick/tock it.

Reply:
Yeah, I have to agree with @sloretz ; while this might currently work for `__slots__`, it seems like something that could break in future versions of Python.   I don't think we should rely on that behavior here.

> Out of all the options, my favorite is making a new API for retrieving members of the message. @clalancette What do you think? Pinging you because it would change a (maybe unintentionally) public API without a good way to tick/tock it.

I'm curious what you have in mind here?  We already have `property`/`setter` methods for the real class that the user interacts with.  Are you suggesting adding a new API to the Metaclass?  And what would that be?

Reply:
> I'm curious what you have in mind here? We already have property/setter methods for the real class that the user interacts with. Are you suggesting adding a new API to the Metaclass? And what would that be?

I was suggesting a new API to get the message fields, followed by replacing any downstream use of `__slots__` with that API. Then we no longer need to modify `__slots__` here as there's already a supported way to get message fields.

Turns out that API already exists 🎉 , so I think we can get rid of modifying `__slots__` altogether.

```python3
>>> from sensor_msgs.msg import JointState
>>> JointState.get_fields_and_field_types()
{'header': 'std_msgs/Header', 'name': 'sequence<string>', 'position': 'sequence<double>', 'velocity': 'sequence<double>', 'effort': 'sequence<double>'}
```

Reply:
I'm not saying we get rid of `__slots__`. I'm saying we don't need to make sure the subclass contains all the message fields. They're already in the `__slots__` of the base class.

Reply:
> I'm not saying we get rid of __slots__. I'm saying we don't need to make sure the subclass contains all the message fields. They're already in the __slots__ of the base class.

Yes!  Thank you for the clear explanation in person, that helped.  Now I understand and I agree.

Reply:
Thanks @almaslov! Just as a comment, on [#194](https://github.com/ros2/rosidl_python/pull/194#issuecomment-1490900345) we encountered the issue of other packages using `__slots__` instead of the method `get_fields_and_field_types()`. I've been working on some PRs on those packages to solve that issue. All the information regarding which packages have been solved or not is on the conversation of #194.